src/graph/executor/query/TraverseExecutor.cpp

```diff
@@ -157,10 +163,98 @@ size_t TraverseExecutor::numRowsOfRpcResp(const RpcResponse& resps) const {
+  auto vidsList = std::make_shared<std::vector<VidHashSet>>(numResps);
+  auto adjLists = std::make_shared<std::vector<VertexMap<Value>>>(numResps);
+  auto taskRunTime = std::make_shared<std::vector<size_t>>(numResps, 0u);
+
+  std::vector<folly::Future<folly::Unit>> futures;
+  futures.reserve(numResps);
+
+  for (size_t i = 0; i < numResps; i++) {
+    auto dataset = resps.responses()[i].get_vertices();
+    if (!dataset) continue;
+    auto func = [this,
+                 dataset = std::move(*dataset),
+                 i,
+                 initVerticesList,
+                 vidsList,
+                 adjLists,
+                 taskRunTime]() {
+      SCOPED_TIMER(&((*taskRunTime)[i]));
+      auto& initVertices = (*initVerticesList)[i];
+      auto& vids = (*vidsList)[i];
+      auto& adjList = (*adjLists)[i];
+      for (GetNbrsRespDataSetIter iter(&dataset); iter.valid(); iter.next()) {
+        Value v = iter.getVertex();
+        initVertices.emplace_back(v);
+        VidHashSet dstSet;
+        auto adjEdges = iter.getAdjEdges(&dstSet);
+        for (const Value& dst : dstSet) {
+          if (adjList_.find(dst) == adjList_.end()) {
+            vids.emplace(dst);
+          }
+        }
+        DCHECK(adjList_.find(v) == adjList_.end())
+            << "The adjacency list should not contain the source vertex: " << v;
+        adjList.emplace(v, std::move(adjEdges));
+      }
+    };
+    futures.emplace_back(folly::via(runner(), std::move(func)));
+  }
+
+  return folly::collect(futures).via(runner()).thenValue(
+      [this, initVerticesList, vidsList, adjLists, taskRunTime](std::vector<folly::Unit>&&) {
+        time::Duration postTaskTime;
+        initVertices_.reserve(sizeOf(*initVerticesList));
+        for (auto& initVertices : *initVerticesList) {
+          std::move(initVertices.begin(), initVertices.end(), std::back_inserter(initVertices_));
+        }
 
+        vids_.reserve(sizeOf(*vidsList));
+        for (auto& vids : *vidsList) {
+          for (auto& v : vids) {
+            vids_.emplace(std::move(v));
+          }
+        }
+
+        adjList_.reserve(adjList_.size() + sizeOf(*adjLists));
+        for (auto& adjList : *adjLists) {
+          for (auto& p : adjList) {
+            adjList_.emplace(std::move(p.first), std::move(p.second));
+          }
+        }
+
+        auto t = postTaskTime.elapsedInUSec();
+        otherStats_.emplace("expandPostTaskTime", folly::sformat("{}(us)", t));
+        folly::dynamic taskRunTimeArray = folly::dynamic::array();
+        for (auto time : *taskRunTime) {
+          taskRunTimeArray.push_back(time);
+        }
+        otherStats_.emplace("expandTaskRunTime", folly::toPrettyJson(taskRunTimeArray));
+
+        return Status::OK();
+      });
+}
+
+folly::Future<Status> TraverseExecutor::expandOneStep(RpcResponse&& resps) {
+  auto numRows = numRowsOfRpcResp(resps);
+  if (numRows < FLAGS_traverse_parallel_threshold_rows) {
```

Comment:
`<` ???

Reply:
When return too many vertices from storage, the collect task will speed a lot time. so not use the parallel execution in that case.